src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -176,35 +187,50 @@ cdef class Buffer:
         if item_offset is None:
             item_offset = 0
         if item_length is None:
-            item_length = size
+            item_length = size - item_offset
         cdef char* cstr = text
         hb_buffer_add_utf8(
             self._hb_buffer, cstr, size, item_offset, item_length)
 
     def add_str(self, text: str,
                 item_offset: int = None, item_length: int = None) -> None:
```

Comment:
The default value for `item_offset` should be 0, and `item_length` should be -1, and then you won’t need any special handling if they were not passed.